[18.0][IMP] l10n_es_aeat_mod347: Field Not in mod347 in partner converted to company dependent

Andrii9090-tecnativa opened the pull request:
This is a straightforward cherry-pick of commit 217854ac76110619a323660dce2e9180b977f8d3

MT-10026 @moduon @rafaelbn @yajo @EmilioPascual @u0f please review if you want :smile:

EmilioPascual committed:
[IMP] l10n_es_aeat_mod347:  Field Not in mod347 in partner converted to company dependent

Issue: https://github.com/OCA/l10n-spain/issues/4173

MT-10026 @moduon

EmilioPascual reviewed (approved):
Gracias @Andrii9090 por la migración de esta funcionalidad, durante la migración en https://github.com/OCA/l10n-spain/pull/4256 se olvido incluir este commit.

Activado en una compañía
<img width="1251" height="725" alt="image" src="https://github.com/user-attachments/assets/5e62c82c-92ee-4763-88fe-80d98195e49e" />

Desactivado en una compañía distinta
<img width="1256" height="717" alt="image" src="https://github.com/user-attachments/assets/9d1b3d6d-127e-4113-a12e-e48caa9fdc06" />

ArantxaSudon reviewed (approved):
Revisado Funcionalmente

Un contacto que marcas como no incluir en el 347 en una compañía en las otras no queda marcado

Gracias @EmilioPascual @Andrii9090

pedrobaeza commented on l10n_es_aeat_mod347/migrations/18.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,13 @@
+
+def migrate(cr, version):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    openupgrade.convert_to_company_dependent(
```
Esto debe detectar si viene de 16.0, donde ya está convertido a dependiente de la compañía, y entonces daría error.

EmilioPascual replied:
Correcto.
@Andrii9090 debes usar `parse_version` importandolo desde `odoo.tools`
Algo parecido a:
```
from odoo.tools import parse_version

...

def migrate(cr, version):
      if parse_version(version) == parse_version(16.0):
            return
      ....
```

Además, debería también comprobarse en el `pre-migration`

pedrobaeza replied:
Bueno, mejor que con eso de la versión, que además, puedes venir de la 17, no de la 16, se debería comprobar por datos. Por ejemplo, si existe la columna en la BD.

Andrii9090-tecnativa replied:
Hemos mejorado el script de migración. Puedes @pedrobaeza  revisar de nuevo, por favor.

pedrobaeza reviewed (changes requested):
Ver comentario inline.

pedrobaeza commented:
Faltaría también el fw-port de la 17, ¿no?

EmilioPascual commented:
> Faltaría también el fw-port de la 17, ¿no?

Se intentará, pero a veces el tiempo no da para más y al final toca priorizar.

pedrobaeza commented:
El fw-port de la 17 debe ser igual que el de la 18, y ésta no se puede aceptar sin tener la anterior.

Xino61122 reviewed (approved):
Revisado Funcionalmente

Un contacto que marcas como no incluir en el 347 en una compañía en las otras no queda marcado 👌.

u0f reviewed (approved):
Gracias @Andrii9090 por migrar esta funcionalidad :smile:. Testeado funcionalmente y todo correcto.

<img width="1920" height="927" alt="image" src="https://github.com/user-attachments/assets/35ea0c38-bcb0-43f6-b4e7-ac81221253a9" />

sabrinaRMartin commented:
Gracias @Andrii9090 !! 
Respecto al código a parte de la sugerencia de Emilio lo veo todo bien.
Lo he probado a nivel funcional también y funciona perfectamente.

yajo committed:
[FIX] l10n_es_aeat_mod347: resilient multi-company field migration

Previous migration script didn't work for v18, where company-dependent fields
are JSONB columns. This commit does the correct migration, and it is only done
if not already done in previous versions.

@moduon MT-10026 MT-10479

pedrobaeza commented on l10n_es_aeat_mod347/migrations/18.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,30 @@
+        "res_partner",
+        "not_in_mod347",
+        "jsonb",
+        SQL(f"jsonb_build_object({args})", *company_ids),
```
No tengo claro que esto vaya a funcionar. ¿Lo habéis probado? ¿Por qué no usar la función de `openupgradelib`?

EmilioPascual replied:
Probado en una migración de 16 a 18 y todo correcto

pedrobaeza replied:
Y por qué no se usa la función de `openupgradelib` que hay para ello?

EmilioPascual replied:
Hola @pedrobaeza, gracias por la sugerencia. En este caso la migración se ha implementado utilizando únicamente métodos del core de Odoo y funciona correctamente así. Que yo sepa no existe ninguna obligatoriedad de usar openupgradelib para realizar migraciones, por lo que preferimos mantener este enfoque.

¡Gracias igualmente por tu revisión y aportaciones!

pedrobaeza replied:
Bueno, son ganas de echar más tiempo teniendo una función que lo hace directamente, pero vosotros mismos, y dado que el mantenimiento en esto es limitado, lo importante es que funcione.

pedrobaeza reviewed (approved):
/ocabot merge nobump
